### PR TITLE
Ledger integration

### PR DESCRIPTION
**Pull request by nc6**
WIP at the moment, submitted to validate the approach.

This adds the following:

- Includes `cardano-chain` in the cabal.project file

- Extends `Ledger.Abstract` with an addition `LedgerConfig` type, used to
  provide static configuration to the ledger update calls.

- Adds a new class `LedgerConfigView`, which must be implemented to allow
  extraction of the ledger configuration from the node config for a given
  protocol.

- Updates the mock ledger to handle these changes.

- Includes a new Byron ledger, which calls through to `updateBody` to update the
  chain state.

### CODE REVIEW

File: ouroboros-consensus/src/Ouroboros/Consensus/Ledger/Byron.hs [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+instance Measured BlockMeasure ByronBlock where
+  measure = blockMeasure
+
+instance HasHeader ByronBlock where
+  type HeaderHash ByronBlock = CC.Block.HeaderHash
+
+  blockHash = CC.Block.blockHash . fmap (const ()) . unByronBlock
+  -- TODO distinguish the genesis hash? How do we do this after the fact?
+  blockPrevHash = BlockHash . CC.Block.blockPrevHash . unByronBlock
+  blockSlot = fromIntegral @Word64 . coerce . CC.Slot.flattenSlotId byronEpochSlots . CC.Block.blockSlot . unByronBlock
```
Comment (erikd) [inline ouroboros-consensus/src/Ouroboros/Consensus/Ledger/Byron.hs]: Isn't `fromIntegral @Word64 . coerce` actually the `SlotNo` constructor?

Comment (nc6) [reply ouroboros-consensus/src/Ouroboros/Consensus/Ledger/Byron.hs]: Hmm, no, needs to unwrap the `FlatSlotId` as well.

File: ouroboros-consensus/src/Ouroboros/Consensus/ChainSyncClient.hs [hunk 2]
```diff
@@ -262,76 +263,85 @@ chainSyncClient _tracer cfg btime (ClockSkew maxSkew) getCurrentChain
...
-      -- TODO #472
-      curLedger <- ledgerState <$> getCurrentLedger
-
       CandidateState {..} <- readTVar varCandidate
 
-      candidateHeaderState' <-
-        case runExcept $ advanceHeader curLedger hdr candidateHeaderState of
-          Left ledgerError            -> disconnect $ LedgerError ledgerError
-          Right candidateHeaderState' -> return candidateHeaderState'
+      ledgerView <-
```
Comment (mrBliss) [inline ouroboros-consensus/src/Ouroboros/Consensus/ChainSyncClient.hs]: Would it make sense to maintain this `ledgerView` in the `CandidateState`? We can then use `atSlot`to check whether it is still valid when rolling forward.

File: ouroboros-consensus/src/Ouroboros/Consensus/ChainSyncClient.hs [hunk 3]
```diff
@@ -262,76 +263,85 @@ chainSyncClient _tracer cfg btime (ClockSkew maxSkew) getCurrentChain
 
       -- Get the 'HeaderState' at genesis (0).
       let candidateChain' = Empty genesisPoint
-          candidateHeaderState' =
-            getHeaderStateFor curLedger candidateChain candidateChain'
+
+      candidateChainState' <- case rewindChainState cfg curChainState (SlotNo 0) of
+        Nothing -> disconnect $ ForkTooDeep genesisPoint theirHead
```
[the remark this answers is not in the document]
Comment (nc6) [reply ouroboros-consensus/src/Ouroboros/Consensus/ChainSyncClient.hs]: I think _not_, but wasn't totally convinced enough to just stick a `Just` in the binding.

[the remark this answers is not in the document]
Comment (nc6) [reply ouroboros-consensus/src/Ouroboros/Consensus/ChainSyncClient.hs]: Internal violation?

File: ouroboros-consensus/src/Ouroboros/Consensus/ChainSyncClient.hs [hunk 4]
```diff
@@ -262,107 +264,95 @@ chainSyncClient _tracer cfg btime (ClockSkew maxSkew) getCurrentChain
...
 
-      candidateHeaderState' <-
-        case runExcept $ advanceHeader curLedger hdr candidateHeaderState of
-          Left ledgerError            -> disconnect $ LedgerError ledgerError
-          Right candidateHeaderState' -> return candidateHeaderState'
+      ledgerView <-
+        case
+          atSlot theirSlot =<< anachronisticProtocolLedgerView cfg curLedger theirSlot
+        of
+          Nothing -> disconnect $ TooFarInTheFuture theirSlot currentSlot
```
Comment (nc6) [inline ouroboros-consensus/src/Ouroboros/Consensus/ChainSyncClient.hs]: Should pause in this situation? We can't go fast enough with block validation to get an up-to-date ledger view, even though we're okay in terms of chain time.

File: ouroboros-consensus/src/Ouroboros/Consensus/Crypto/DSIGN/Cardano.hs [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+import           GHC.Generics (Generic)
+
+import           Ouroboros.Consensus.Crypto.DSIGN.Class
+import           Ouroboros.Consensus.Util.Condense
+
+pm :: ProtocolMagicId
+pm = undefined
+
+st :: SignTag
+st = undefined
```
Comment (nc6) [inline ouroboros-consensus/src/Ouroboros/Consensus/Crypto/DSIGN/Cardano.hs]: Associated constraint on a

**Comment by nc6** [comment]
Closed in favour of #531

**Pull request closed without merging**